src/dev/flang/ast/AbstractType.java

```diff
@@ -1562,6 +1564,64 @@ static boolean checkActualTypePars(SourcePosition pos, AbstractFeature called, L
     return result;
   }
 
+
+  /**
+   * collect all the directly and indirectly - via generics -
+   * used types in this type.
+   * example: Cons i32 (list i32) => { Cons, i32, list }
+   *
+   * @return
+   */
+  private Set<AbstractType> usedTypes()
+  {
+    var result = new TreeSet<AbstractType>();
+    usedTypes(result);
+    return result;
+  }
+
+
+  /**
+   * recursion helper for `usedTypes (no args)`
+   * @param result
+   */
+  private void usedTypes(TreeSet<AbstractType> result)
+  {
+    if (result.add(this))
+      {
+        for (AbstractType at : this.generics())
+          {
+            at.usedTypes(result);
+          }
+        if (isChoice())
+          {
+            for (AbstractType at : this.choiceGenerics())
+              {
+                at.usedTypes(result);
+              }
+          }
+      }
+  }
+
+
+  /**
+   * This type and any of its generics that are less visible than `v`.
+   *
+   * @param v
+   * @return
+   */
+  public Set<AbstractType> lessVisibleThan(Visi v)
+  {
+    if (PRECONDITIONS) require
+      (!v.definesTypeVisibility());
+
+    return isGenericArgument()
+      ? genericArgument().constraint().lessVisibleThan(v)
```

Comment:
The visibiliy of type parameter should probably not depend on the constraint, but on its underlying feature, i.e., `genericArgument().typeParameter()`?

Reply:
now, done as suggested